internal/transports/p2p/peer/peer.go

```diff
@@ -0,0 +1,558 @@
+
+	peer := &Peer{
+		addr:            netAddr,
+		cfg:             cfg,
+		chainParams:     chainParams,
+		headersService:  headersService,
+		chainService:    chainService,
+		checkpoints:     chainParams.Checkpoints,
+		nextCheckpoint:  nextCheckpoint,
+		log:             log,
+		services:        wire.SFspv,
+		protocolVersion: initialProtocolVersion,
+		synced:          nextCheckpoint == nil,
+		wg:              sync.WaitGroup{},
+		msgChan:         make(chan wire.Message),
+		quitting:        false,
+		quit:            make(chan struct{}),
+	}
+	return peer, nil
+}
+
+func (p *Peer) Connect() error {
+	conn, err := net.Dial(p.addr.Network(), p.addr.String())
+	if err != nil {
+		p.log.Error().Msgf("error connecting to the peer, reason: %v", err)
+		return err
+	}
+	p.conn = conn
+	p.log.Info().Msgf("connected to peer: %s", p)
+	return nil
+}
+
+func (p *Peer) Disconnect() error {
+	p.log.Info().Msgf("disconnecting peer: %s", p)
+
+	p.quitting = true
+	close(p.quit)
+	err := p.conn.Close()
+	if err != nil {
+		return err
+	}
+
+	p.wg.Wait()
+	p.log.Info().Msgf("successfully disconnected peer %s", p)
+	return nil
+}
+
+func (p *Peer) Start() error {
+	err := p.negotiateProtocol()
+	if err != nil {
+		p.log.Error().Msgf("error negotiating protocol with peer %s, reason: %v", p, err)
+		return err
+	}
+
+	go p.writeMsgHandler()
+	go p.readMsgHandler()
+	go p.pingHandler()
+
+	err = p.startSync()
+	if err != nil {
+		p.log.Error().Msgf("error starting sync from peer %s, reason: %v", p, err)
+		return err
+	}
+	return nil
+}
+
+func (p *Peer) startSync() error {
+	p.log.Info().Msgf("requesting headers from peer %s", p)
+
+	var err error
+	if p.nextCheckpoint != nil {
+		err = p.writeGetHeadersMsg(p.nextCheckpoint.Hash)
+	} else {
+		err = p.writeGetHeadersMsg(&zeroHash)
+	}
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Peer) negotiateProtocol() error {
+	err := p.writeOurVersionMsg()
+	if err != nil {
+		return err
+	}
+	p.log.Info().Msgf("version sent to peer: %s", p)
+
+	err = p.readVerOrVerAck()
+	if err != nil {
+		return err
+	}
+
+	err = p.readVerOrVerAck()
+	if err != nil {
+		return err
+	}
+	p.log.Info().Msgf("received version and verack from peer: %s", p)
+
+	err = p.writeMessage(wire.NewMsgVerAck())
+	if err != nil {
+		return err
+	}
+
+	p.log.Info().Msgf("protocol negotiated successfully with peer: %s", p)
+	return nil
+}
+
+// PingHandler is a handler for sending ping messages to peers.
+// Must be run as a goroutine.
+func (p *Peer) pingHandler() {
+	p.wg.Add(1)
+	pingTicker := time.NewTicker(pingInterval)
+	defer pingTicker.Stop()
+
+out:
+	for {
+		select {
+		case <-pingTicker.C:
+			nonce, err := wire.RandomUint64()
+			if err != nil {
+				p.log.Error().Msgf("error generating nonce for ping msg, reason: %v", err)
+				continue
+			}
+			p.log.Info().Msgf("sending ping to peer %s with nonce: %d", p, nonce)
+			p.queueMessage(wire.NewMsgPing(nonce))
+
+		case <-p.quit:
+			break out
```

Comment:
Consider not using the label "out". 
Instead you can just put here 
```golang
p.log.Info().Msgf("....")
return
```

and after `p.wg.Add` at the begining of the function, add:
```golang
defer p.wg.Done()
```

Reply:
Absolutely,  done :)